[JSC] Implement DIV2BY1 division algorithm

Constellation opened the pull request:
#### b916e2ba4eea871087af542481ca559c8723613c
<pre>
[JSC] Implement DIV2BY1 division algorithm
<a href="https://bugs.webkit.org/show_bug.cgi?id=303964">https://bugs.webkit.org/show_bug.cgi?id=303964</a>
<a href="https://rdar.apple.com/166268064">rdar://166268064</a>

Reviewed by Sosuke Suzuki.

Implement &quot;Division by Invariant Integers using Multiplication&quot;&apos;s DIV2BY1 algorithm,
which is widely used in various compilers and BigInt implementations.
This performs division faster by using pre-computed inverse number of
the divisor. This is effective when we repeatedly divides numbers with
the same divisor. This is common in our JSBigInt division / remainder
algorithm.

We also remove inplaceSub etc. case when qhat is 0. qhatvSpan becomes
complete zeros, and inplaceSub etc. has zero effect.

* Source/JavaScriptCore/runtime/JSBigInt.cpp:
(JSC::DigitDiv::DigitDiv):
(JSC::DigitDiv::div):
(JSC::DigitDiv::calculateInverse):
(JSC::JSBigInt::divideTextbook):
* Source/JavaScriptCore/runtime/JSBigInt.h:

Canonical link: <a href="https://commits.webkit.org/304278@main">https://commits.webkit.org/304278@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eb55735fc375575bbc4cdb060f7c95e6773430cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135114 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7506 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46370 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142620 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8141 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7355 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103246 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138060 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5788 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121097 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84102 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/5592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/3204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3214 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/127124 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114806 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39271 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145318 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133606 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7190 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39842 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111618 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7234 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6025 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111985 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5433 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117393 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61117 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20841 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7243 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35543 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166475 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7002 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70800 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7226 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7105 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSBigInt.h:
```diff
@@ -485,6 +485,17 @@ class JSBigInt final : public JSCell {
     static constexpr unsigned halfDigitBits = digitBits / 2;
     static constexpr Digit halfDigitMask = (1ull << halfDigitBits) - 1;
 
+    // Digit arithmetic helpers.
+    static Digit digitAdd(Digit a, Digit b, Digit& carry);
+    static Digit digitAdd3(Digit a, Digit b, Digit c, Digit& carry);
+    static Digit digitSub(Digit a, Digit b, Digit& borrow);
+    static Digit digitSub2(Digit a, Digit b, Digit borrowIn, Digit& borrowOut);
+    static std::tuple<Digit, Digit> digitMul(Digit a, Digit b);
+    static Digit digitDiv(Digit high, Digit low, Digit divisor, Digit& remainder);
+    static Digit digitPow(Digit base, Digit exponent);
+    static Digit subtractAndReturnBorrow(std::span<Digit> z, std::span<const Digit> x, std::span<const Digit> y);
+    static Digit addAndReturnCarry(std::span<Digit> z, std::span<const Digit> x, std::span<const Digit> y);
```
Should we make `DigitDiv` friend class of `BigInt instead of making these methods public?

Constellation replied:
Since these digit arithmetics are in general useful, I would like to keep it public for random use.

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSBigInt.cpp:
```diff
@@ -1136,7 +1207,7 @@ std::tuple<std::span<JSBigInt::Digit>, std::span<JSBigInt::Digit>> JSBigInt::div
             // Estimate the current quotient digit by dividing the most significant
             // digits of dividend and divisor. The result will not be too small,
             // but could be a bit too large.
-            qhat = digitDiv(ujn, uSpan[j + n - 1], vn1, rhat);
```
I believe `divq` is used in x86_64, so this might be faster than the C++ implementation added by this change? However, this is just a guess, so could you create a benchmark?

Constellation replied:
In fact still `divq` is slower than the C++ implementation because even hardware implementation is taking more cycles than the algorithm used here. The key of this algorithm is computing the inverse value a-priori by leveraging the fact that we are using the same divisor for many iterations.

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSBigInt.cpp:
```diff
@@ -932,6 +932,76 @@ JSValue JSBigInt::multiply(JSGlobalObject* globalObject, JSBigInt* x, int32_t y)
+        Digit refR = 0;
+        Digit refQ = JSBigInt::digitDiv(u1, u0, d, refR);
+        if (!(refR == r && refQ == q)) [[unlikely]] {
+            dataLogLn(u1, " ", u0, " ", d, " ", refR, " ", r, " ", refR == r, " ", refQ, " ", q, " ", refQ == q);
```
Is this `dataLogLn` intentionally kept?

Constellation replied:
Yes, this is assertion which fails. So dumping values for debuggability. (And compiled only when ASSERT_ENABLED is true).

sosukesuzuki reviewed (approved):
r=me, please fix the style issue if it is necessary

Constellation committed:
[JSC] Implement DIV2BY1 division algorithm
https://bugs.webkit.org/show_bug.cgi?id=303964
rdar://166268064

Reviewed by Sosuke Suzuki.

Implement "Division by Invariant Integers using Multiplication"'s DIV2BY1 algorithm,
which is widely used in various compilers and BigInt implementations.
This performs division faster by using pre-computed inverse number of
the divisor. This is effective when we repeatedly divides numbers with
the same divisor. This is common in our JSBigInt division / remainder
algorithm.

We also remove inplaceSub etc. case when qhat is 0. qhatvSpan becomes
complete zeros, and inplaceSub etc. has zero effect.

* Source/JavaScriptCore/runtime/JSBigInt.cpp:
(JSC::DigitDiv::DigitDiv):
(JSC::DigitDiv::div):
(JSC::DigitDiv::calculateInverse):
(JSC::JSBigInt::divideTextbook):
* Source/JavaScriptCore/runtime/JSBigInt.h:

Canonical link: https://commits.webkit.org/304278@main
